### PR TITLE
[druid] Allow custom druid postaggregators

### PR DESCRIPTION
**Pull request by brianwolfe**
Fix the postaggregation for approxHistogram quantiles so it adds
the dependent field and that can show up in the graphs/tables.

In general, postAggregators add significant power, in developing custom
postAggregators, I've run into barriers integrating with Superset. Plywood
and other Druid query planners typically have escape hatches to specify
custom postAggregators without requiring a custom type definition for each
one.

This commit adds a similar capability for Superset and a additional
field/fields/fieldName breakdown of the typical naming for dependent
aggregations, which should make it significantly easier to develop
approxHistogram and custom postAggregation-required dashboards.

This should be able to supercede:
https://github.com/apache/incubator-superset/pull/2345
since it is strictly more general.

This should at least pull in the right dependent metrics and allow some custom aggregations and quantile/histogram links. @mistercrunch

### CODE REVIEW

**Comment by xrmx** [comment]
Could please fix tests with python2.7 and remove all the other unrelated changes you did in the PR please?

**Comment by brianwolfe** [comment]
Not sure what happened with that, it wasn't in my commit history. Fixed via rebase, will fix remaining python2 issues shortly, when I can get duplication on my local machine.

**Comment by brianwolfe** [comment]
@xrmx, sorry for the busted PR earlier. This should be good to go (assuming my local testing and Travis agree).

I pulled the helper function out so I could actually unit test the dependent-data extraction/naive query planning. It'll be a significantly larger scope of testing changes if I put that back.

**Comment by brianwolfe** [comment]
Attempting to rerun travis to get coveralls passing.

**Comment by brianwolfe** [comment]
@xrmx or @mistercrunch PTAL This PR has passed Travis https://travis-ci.org/apache/incubator-superset/builds/254976199 (aside from the untouched JS build) and coveralls looks good: https://coveralls.io/builds/12486731

**Comment by brianwolfe** [comment]
Resetting travis build again. Hopefully this all passes so that this can be merged in.

File: superset/connectors/druid/models.py [hunk 1]
```diff
@@ -718,51 +718,29 @@ def values_for_column(self,
...
         all_metrics = []
         post_aggs = {}
 
-        columns_dict = {c.column_name: c for c in self.columns}
-
         def recursive_get_fields(_conf):
-            _fields = _conf.get('fields', [])
+            print(_conf)
+            _type = _conf.get('type')
+            _field = _conf.get('field', None)
```
Comment (mistercrunch) [inline superset/connectors/druid/models.py]: `, None` is implicit, that's the default behavior

File: superset/connectors/druid/models.py [hunk 2]
```diff
@@ -799,11 +777,54 @@ def recursive_get_fields(_conf):
...
                     )
-                else:
+                elif mconf.get('type') == 'arithmetic':
                     post_aggs[metric_name] = Postaggregator(
                         mconf.get('fn', "/"),
                         mconf.get('fields', []),
                         mconf.get('name', ''))
+                else:
+                    post_aggs[metric_name] = Postaggregator(
+                        None,
```
Comment (mistercrunch) [inline superset/connectors/druid/models.py]: it would help readability to use keyword args here

File: superset/connectors/druid/models.py [hunk 3]
```diff
@@ -799,11 +777,54 @@ def recursive_get_fields(_conf):
...
+                        None,
+                        mconf.get('name', ''))
+                    post_aggs[metric_name].post_aggregator = mconf
+        return all_metrics, post_aggs
+
+    def run_query(  # noqa / druid
```
Comment (mistercrunch) [inline superset/connectors/druid/models.py]: I'm not sure why git is getting the diffing all wrong here, things are not lining up. Makes it hard to see what has changed in this section. Maybe moving `_metrics_and_post_aggs` higher / lower in the class definition might help git diffing it right?

File: superset/connectors/druid/models.py [hunk 4]
```diff
@@ -799,11 +777,54 @@ def recursive_get_fields(_conf):
...
                         mconf.get('fn', "/"),
                         mconf.get('fields', []),
                         mconf.get('name', ''))
+                else:
+                    post_aggs[metric_name] = Postaggregator(
+                        None,
+                        None,
+                        mconf.get('name', ''))
+                    post_aggs[metric_name].post_aggregator = mconf
+        return all_metrics, post_aggs
```
Comment (mistercrunch) [inline superset/connectors/druid/models.py]: I'm never a big fan of returning tuples but I think it's ok in this case

File: superset/connectors/druid/models.py [hunk 5]
```diff
@@ -718,51 +718,29 @@ def values_for_column(self,
...
-            for _f in _fields:
-                _type = _f.get('type')
-                if _type in ['fieldAccess', 'hyperUniqueCardinality']:
-                    field_names.append(_f.get('fieldName'))
-                elif _type == 'arithmetic':
+            if _type in ['fieldAccess', 'hyperUniqueCardinality',
+                         'quantile', 'quantiles']:
+                field_names.append(_conf.get('fieldName', ''))
+
+            if _field is not None:
```
Comment (mistercrunch) [inline superset/connectors/druid/models.py]: `if _field:` is the pythonesque convention

File: superset/connectors/druid/models.py [hunk 6]
```diff
@@ -718,51 +718,29 @@ def values_for_column(self,
...
+                field_names += recursive_get_fields(_field)
+
+            if _fields is not None:
```
Comment (mistercrunch) [inline superset/connectors/druid/models.py]: why not use the same if block as the one right above this one?

**Review by mistercrunch — CHANGES_REQUESTED** [review]
This is neat, thanks for the PR! 

So can post aggregators build upon other post aggregators? Seems like it's supported here and want to make sure Druid allows for that.

File: superset/connectors/druid/models.py (hunk 6 above)
Comment (brianwolfe) [inline superset/connectors/druid/models.py]: _field vs _fields.

File: superset/connectors/druid/models.py (hunk 5 above)
Comment (brianwolfe) [inline superset/connectors/druid/models.py]: I suppose. I prefer to be explicit when I care about None vs. when I want to reject all falsy values. https://docs.python.org/2.4/lib/truth.html

I'll change the PR.

Comment (brianwolfe) [inline superset/connectors/druid/models.py]: https://www.python.org/dev/peps/pep-0008/#programming-recommendations

**Comment by brianwolfe** [comment]
@mistercrunch Yes, postAggregators can reference other postAggregators. The dataSketches library is rather incredible if you haven't looked at it before: https://datasketches.github.io/

They combine PostAggregators extensively to compute interesting results.

https://github.com/druid-io/druid/blob/master/extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchSetPostAggregator.java#L49

Typically they are combined by FieldReference, but not always. In general, we'd probably want a real query planner here, but this allows us to use approxHistograms, dataSketches and other libraries without friction for now.

File: superset/connectors/druid/models.py (hunk 6 above)
Comment (brianwolfe) [inline superset/connectors/druid/models.py]: this is mostly a matter of supporting a newer and older postAggregators that have different customs for storing the dependent fields.

File: superset/connectors/druid/models.py (hunk 2 above)
Comment (brianwolfe) [inline superset/connectors/druid/models.py]: fixed by making a CustomPostAggregator class that encompasses this logic. Should make more sense that way.

File: superset/connectors/druid/models.py (hunk 4 above)
Comment (brianwolfe) [inline superset/connectors/druid/models.py]: yeah, for an internal api, I feel like it's kinda defensible.

File: superset/connectors/druid/models.py (hunk 3 above)
Comment (brianwolfe) [inline superset/connectors/druid/models.py]: moved up to other static methods.

**Comment by mistercrunch** [comment]
This LGTM at this point. We'll need to validate in staging that existing slices that make use of postaggregators still compute well.

**Pull request merged**